circular.py

```diff
@@ -10,9 +15,79 @@ def _import_from_path(module_name, file_path):
     spec.loader.exec_module(module)
 
 
-def test_circular_imports():
-    for path, _, files in os.walk("starknet_py"):
+def assert_no_circular_imports():
+    for path, _, files in os.walk(PACKAGE_NAME):
         py_files = [f for f in files if f.endswith(".py")]
-        for file_ in py_files:
-            file_path = os.path.join(path, file_)
-            _import_from_path(file_, file_path)
+        for file in py_files:
+            file_path = os.path.join(path, file)
+
+            relative_path = os.path.relpath(file_path, PACKAGE_NAME)
+            module_path_no_ext = relative_path.removesuffix(".py")
+            # Handle __init__.py files specially
+            if module_path_no_ext.endswith("__init__"):
+                module_path_no_init = module_path_no_ext.removesuffix(
+                    "__init__"
+                ).rstrip(os.sep)
+                if not module_path_no_init:  # Top-level __init__.py
+                    module_name = PACKAGE_NAME
+                else:
+                    dotted_module_path = module_path_no_init.replace(os.sep, ".")
+                    module_name = f"{PACKAGE_NAME}.{dotted_module_path}"
+            else:
+                dotted_module_path = module_path_no_ext.replace(os.sep, ".")
+                module_name = f"{PACKAGE_NAME}.{dotted_module_path}"
+
+            _import_from_path(module_name, file_path)
+
+
+def test_circular_imports_absent():
+    assert_no_circular_imports()
+
+
+def _run_circular_import_test(module_name, import_a, import_b):
+    module_path = os.path.join(PACKAGE_NAME, module_name)
+    os.makedirs(module_path, exist_ok=True)
+    try:
+        with open(os.path.join(module_path, "__init__.py"), "w") as f:
+            f.write("")
+        with open(os.path.join(module_path, "file_a.py"), "w") as f:
+            f.write(f"{import_a}\nclass A:\n    pass\n")
+        with open(os.path.join(module_path, "file_b.py"), "w") as f:
+            f.write(f"{import_b}\nclass B:\n    pass\n")
+        error_regex = (
+            rf"(?:"
+            rf"cannot import name 'A' from '{PACKAGE_NAME}.{module_name}.file_a' \(.*{PACKAGE_NAME}[\\/]+{module_name}[\\/]+file_a\.py\)"
+            rf"|"
+            rf"cannot import name 'B' from '{PACKAGE_NAME}.{module_name}.file_b' \(.*{PACKAGE_NAME}[\\/]+{module_name}[\\/]+file_b\.py\)"
+            rf")"
+        )
+        with pytest.raises(ImportError, match=error_regex):
+            assert_no_circular_imports()
+    finally:
+        # Clean up temporary files
+        if os.path.exists(module_path):
+            shutil.rmtree(module_path)
+        sys.modules.pop(f"{PACKAGE_NAME}.{module_name}.file_a", None)
+        sys.modules.pop(f"{PACKAGE_NAME}.{module_name}.file_b", None)
+        sys.modules.pop(f"{PACKAGE_NAME}.{module_name}", None)
```

Comment:
Wouldn't it be easier to maintain if we created the two packages you are using for two tests below and kept them somewhere in the repo? And run tests on them? Instead of maintaining this logic.

Reply:
We can't have these files in other package than starknet_py because we cannot have multiple top-level packages in a flat layout, e.g. https://app.readthedocs.org/projects/starknetpy/builds/30467365/